### ISSUE TITLE
Using user timezone on each render/motion process (around_ callback?)

### ISSUE THREAD

**Issue by davidalejandroaguilar** [unabridged/motion#20]
# Feature Request

TL;DR I think we're missing an around_ callback around rendering.

# Description

I have a component that displays time based on the user's timezone. To achieve this on initial request, I use the user's timezone for that request:

```rb
class ApplicationController < ActionController::Base
  around_action :use_user_timezone, if: :current_user

  private

  def use_user_timezone(&block)
    Time.use_zone(current_user.timezone, &block)
  end
end
```

However, when Motion re-renders my component, it does so with UTC time.

I have tried:

* Using `Time.use_zone(current_user.timezone) do` around the component mapped motion so that everything database based is done with the user's timezone. However, this only affects backend stuff and not the actual re-rendering. Also, it is very cumbersome to wrap every method by hand.

```rb
class TaskComponent < ViewComponent::Base
  include Motion::Component

  map_motion :do_something

  def do_something
    Time.use_zone(current_user.timezone) do
      object.perform
    end
  end
end
```

* Setting timezone on `before_render`. This works but it will not unset it after rendering is done.

```rb
class TaskComponent < ViewComponent::Base
  include Motion::Component

  map_motion :do_something

  def do_something
    object.perform
  end

  def before_render
    Time.zone = current_user.zone
  end
end
```

* Setting timezone on an ActionCable callback. Couldn't make this work.

I think what we're missing here is something akin to an `around_` callback that we can hook into to set these kind of things. I'm thinking this could be added to `process_motion` but I'm not really familiar with the codebase so I'm open to ideas.

# Reproduce the issue

I have created a demo repo [here](https://github.com/davidalejandroaguilar/motion_timezone), it has a sqlite db so I think it should be ready to go. Else, just create a post. Then, click the "Update" button. You can see the timezone changing to UTC from Guadalajara on every re-render.

![img](https://i.imgur.com/CJ2FMiP.gif)

**Comment by davidalejandroaguilar** [comment]
It's worth mentioning that Stimulus Reflex has [added](https://github.com/hopsoft/stimulus_reflex/blob/master/lib/stimulus_reflex/reflex.rb#L18-L20) these kind of callbacks which I've used for this very same thing.

**Comment by alecdotninja** [comment]
Thank you for the detailed feature request, @davidalejandroaguilar !

A temporary workaround that you can use for this is something like:
```ruby
def render_in(*)
  Time.use_zone(current_user.timezone) { super }
end
```

I agree that this is cumbersome though, and an `around_render` callback is a much better fit.

With the current design, I think the most straight-forward thing to do is to add `before_`/`after_`/`around_` callbacks for [`render`](https://github.com/unabridged/motion/blob/5b43ea7852ec152245966163cd9b1ddde1d2261a/lib/motion/component/rendering.rb#L30), [`motion`](https://github.com/unabridged/motion/blob/5b43ea7852ec152245966163cd9b1ddde1d2261a/lib/motion/component/motions.rb#L45), [`broadcast`](https://github.com/unabridged/motion/blob/5b43ea7852ec152245966163cd9b1ddde1d2261a/lib/motion/component/broadcasts.rb#L75), and [`periodic_timer`](https://github.com/unabridged/motion/blob/5b43ea7852ec152245966163cd9b1ddde1d2261a/lib/motion/component/periodic_timers.rb#L49).

1. I think it will be awkward to also have a `before_render` instance method with `ViewComponent::Base` (but of course removing it is beyond the scope of `Motion`).
1. If you want to have a callback that runs around _everything_, you have to define 4 around callbacks.

@latortuga @Gmfholley I'm interested in your thoughts.

**Comment by alecdotninja** [comment]
Another thought after sleeping on it: We haven't used the word "action" yet. We could `before_action`/`after_action`/`around_around` callbacks that run around _everything_ instead of (or in addition to) the callbacks I suggested above.

**Comment by alecdotninja** [comment]
This feature is now available as `around_action` in `0.3.0`. See #27 for details.
